verilog/analysis/checkers/always_comb_blocking_rule.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright 2017-2019 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/always_comb_blocking_rule.h"
+
+#include <string>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::SyntaxTreeContext;
+
+// Register AlwaysCombBlockingRule
+VERILOG_REGISTER_LINT_RULE(AlwaysCombBlockingRule);
+
+absl::string_view AlwaysCombBlockingRule::Name() { return "always-comb-blocking"; }
+const char AlwaysCombBlockingRule::kTopic[] = "combinational-logic";
+const char AlwaysCombBlockingRule::kMessage[] =
+    "Use blocking assignment instead of non-blocking one.";
+
+std::string AlwaysCombBlockingRule::GetDescription(DescriptionType description_type) {
+  return absl::StrCat("Checks that there are no occurrences of ",
+                      "non-blocking assignment in combinational logic.");
+}
+
+void AlwaysCombBlockingRule::HandleSymbol(const verible::Symbol& symbol,
+                                  const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (always_comb_matcher_.Matches(symbol, &manager)) {
+    violations_.push_back(LintViolation(symbol, kMessage, context));
+  }
+  if (always_comb_block_matcher_.Matches(symbol, &manager)) {
+    violations_.push_back(LintViolation(symbol, kMessage, context));
+  }
```

Comment:
An alternative to consider instead of using a matcher is SearchSyntaxTree.
For example:
https://github.com/google/verible/blob/f74e900dbdd40c48fcaf1ff68db0c7f5bd342658/verilog/CST/module.cc#L33

Once you're inside `always_comb_matcher_` above, you could collect all violating constructs under the matching symbol sub-tree.  This saves you from having to write an elaborate path matcher.

Also instead of reporting a violation over the whole matched symbol, I would loop over the search matches and only report the ranges of the offending subnodes (mismatched assignment style).  Either the whole assignment statement or even the assignment operator would be appropriate.

Reply:
The idea with reporting assignment operator fitted me the most.